### PR TITLE
Run TLS destructors for main thread on UNIX platforms

### PR DESCRIPTION
**Pull request by surban**
This calls TLS destructors on UNIX platforms other than Linux when the main thread exits. This is done by registering a process-wide atexit handler.

r? joboet

### CODE REVIEW

**Comment by joboet** [comment]
Hi, T-libs! Are you okay with this behaviour change? This makes platforms like musl behave the same way as glibc; they will now run TLS destructors both when `main` returns and when `exit` is called instead of only running them when a thread exits.

@rustbot label +I-libs-nominated

File: library/std/src/sys/thread_local/guard/key.rs [hunk 1]
```diff
@@ -7,17 +7,38 @@ use crate::sys::thread_local::key::{LazyKey, set};
...
+    fn enable_process() {
+        use crate::sync::atomic::{AtomicBool, Ordering};
+        use crate::sys::thread_local::key::at_process_exit;
 
-    // Setting the key value to something other than NULL will result in the
-    // destructor being run at thread exit.
-    unsafe {
-        set(DTORS.force(), ptr::without_provenance_mut(1));
+        static REGISTERED: AtomicBool = AtomicBool::new(false);
+        if !REGISTERED.swap(true, Ordering::AcqRel) {
```
Comment (m-ou-se) [inline library/std/src/sys/thread_local/guard/key.rs:30]: Relaxed memory ordering is enough for this sort of thing.

**Comment by Amanieu** [comment]
This is currently documented as platform-specific behavior [here](https://doc.rust-lang.org/nightly/std/thread/struct.LocalKey.html#platform-specific-behavior).

The main benefit of this PR would be to make the behavior more consistent across platforms so that we can remove the warning in the documentation and make it a guarantee. However for that to happen we need to be able to ensure that this works on *all* platforms. Could you have a look at other (non-UNIX) platforms to see what the current behavior is and whether they need a similar workaround?

File: library/std/src/sys/thread_local/guard/key.rs (hunk 1 above)
Comment (surban) [reply library/std/src/sys/thread_local/guard/key.rs:30]: My fear of memory models always makes me choose the safest option ;-)

**Comment by surban** [comment]
> However for that to happen we need to be able to ensure that this works on _all_ platforms. Could you have a look at other (non-UNIX) platforms to see what the current behavior is and whether they need a similar workaround?

I can have a look, however I certainly don't have access to all platforms supported by Rust. I can test on Windows, Linux, FreeBSD and WASI preview 1.

Is there a way to run test programs on all Tier 1 and Tier 2 targets via CI, so that I could run automatic tests?

**Comment by surban** [comment]
@Amanieu

So this what I was able to gather by looking at the code and doing some testing where possible.

- Linux (and most likely Android, Fuchsia, Redox, Hurd, NetBSD, Dragonfly) (tier 1)
  - recent glibc with symbol `__cxa_thread_atexit_impl`
    - :white_check_mark: TLS destructors did run for thread exits and process exit 
  - old glibc or musl or other libc without symbol `__cxa_thread_atexit_impl`
    - :+1: TLS destructors did _not_ run for process exit, but _do_ run with this PR
  - tested with glibc and musl libc on Ubuntu 22.04

- Windows (tier 1)
  - :white_check_mark:  TLS destructors did run for thread exits and process exit
  - tested using both MSVC and GNU toolchains on Windows 11 x64

- Apple platforms (tier 1)
  - :white_check_mark:  TLS destructors did run for thread exits and process exit
  - tested using macOS 14.4

- other UNIX (tier 2)
  - :+1: TLS destructors did _not_ run for process exit, but _do_ run with this PR
  - tested using FreeBSD 14.2

- targets without threads (`wasm32-wasip1` without `atomics` feature and UEFI) (tier 2)
  - :x: TLS destructors are not tracked
  - currently there is no bookkeeping of TLS values that need to be dropped
  - would require adding `destructors::register` calls to the `statik` module

- `wasm32-unknown-unknown` with `atomics` feature (tier 2)
  - :no_entry_sign:  has no concept of process and thread exit

- Intel/Fortanix SGX (tier 2)
  -  :no_entry_sign: has no concept of process exit

- Risc0 zkvm (tier 3)
  -  :no_entry_sign: has no concept of process exit

- Xous (tier 3)
  - :no_entry_sign: has no concept of process exit

- Hermit (tier 3)
  - :white_check_mark: calls `thread_local::destructors::run()` explicitly after `main()` and thread exit
  - should work, but not tested

- Solid (tier 3)
  - :question: depends on whether the system calls functions registered through `SOLID_TLS_AddDestructor` at process exit
  - I did not find any API documentation.

### Summary

Targets that do not have support for threads (i.e. `wasm32-wasip1` without `atomics` and UEFI) require the `statik` module to be extended to track TLS values for destruction at process exit. For `wasm32-wasip1` this can be invoked from an `atexit` handler, while UEFI might be more challenging.

The remaining non-UNIX targets should either work (Hermit, Solid) or do not have the concept of process exit (`std::process:exit` is unimplemented).

**Comment by joboet** [comment]
That's a great summary, thank you! One other thing to note though is that not every UNIX platform has native thread local storage, and that the solution proposed in this PR will not work for them since `std` does not manage the destructors itself; this is left to the `pthread` TLS key implementation.

**Comment by surban** [comment]
> One other thing to note though is that not every UNIX platform has native thread local storage, and that the solution proposed in this PR will not work for them since `std` does not manage the destructors itself; this is left to the `pthread` TLS key implementation.

Could you name such a platform so that I can test? Preferably something that can be run in a VM.

**Comment by joboet** [comment]
Oh wow, they've become quite rare since last time I checked. But Solaris (e.g. `x86_64_pc_solaris`) is one for instance.

**Comment by surban** [comment]
Closing in favor of #134085

**Pull request closed without merging**